Реализация задач ТЗ спринта 9

h0ttab opened the pull request:
### В рамках финального задания спринта 9 в приложение были внесены следующие изменения:
## По тех. заданию:
- Добавлен HTTP RESTful API с поддержкой всех необходимых эндпоинтов
- Обновлены старые и добавлены новые unit-тесты

## Дополнительно:
- Добавлено несколько новых классов для исключений
- Реализованы кастомные (де-)сериализаторы для сущностей задач, а также адаптеры типов
- Очень большую часть логики обработчиков и их тестов удалось выделить в параметризованный абстрактный класс
- Добавлено несколько интерфейсов, что позволило сделать структуру классов более организованной
- Все служебные сообщения (консоль + ответы API) были переписаны на английский язык для соответствия общепринятым практикам
- Я начал практиковаться в использовании стандарта Conventional Commits для оформления коммитов
- Был актуализирован файл README.md
- В readme была добавлена инструкция по запуску проекта
- В разделе "release" в репозитории сохранён черновик с релизной версией приложения: сама программа, упакованная в .jar, плюс .har файл для импорта коллекции HTTP запросов в Postman или Insomnia

h0ttab committed:
Импортировать GSON

h0ttab committed:
Создать класс HttpTaskServer

h0ttab committed:
Создать класс для формирования ответа с ошибкой

h0ttab committed:
Создать абстрактный класс BaseHttpHandler

h0ttab committed:
Создать исключение для случаев пересечения задач

h0ttab committed:
Создать исключение NotFoundException

h0ttab committed:
Интегрировать новые исключения в менеджеры

h0ttab committed:
Интегрировать новые исключения в демонстрационные сценарий

h0ttab committed:
Интегрировать новые исключения в TaskManagerTest

h0ttab committed:
Интегрировать новые исключения в FileBackedTaskManagerTest

h0ttab committed:
Интегрировать новые исключения в ManagersTest

h0ttab committed:
Интегрировать новые исключения в EpicTest

h0ttab committed:
Интегрировать новые исключения в SubTaskTest

h0ttab committed:
Адаптировать тесты TaskManagerTest под новый механизм валидации пересечений

h0ttab committed:
Настроить экземпляр Gson, создать адаптеры типов для LocalDateTime и Duration

h0ttab committed:
Рефакторинг

h0ttab committed:
Обработка GET методов для пути /tasks

h0ttab committed:
Добавить интерфейс для CRUD операция с тасками

h0ttab committed:
Рефакторинг

h0ttab committed:
Добавление новых шаблонных ответов и рефакторинг

h0ttab committed:
Добавить маршрутизацию по пути /tasks

h0ttab committed:
Создать обработчик для /tasks

h0ttab committed:
Создать enum для названий типов задач

h0ttab committed:
Выделить логику отправки ответов в отдельный класс со static методами

h0ttab committed:
Выделить методы CRUD операций обработчиков в отдельный параметризованный интерфейс

h0ttab committed:
Выделить общую для всех обработчиков логику в параметризованный абстрактный класс

h0ttab committed:
Обновить TasksHttpHandler под новую реализацию абстрактного обработчика

h0ttab committed:
Рефакторинг

h0ttab committed:
Создать сериализатор/десериализатор для гарантированно корректной работы JSON с Epic

h0ttab committed:
Реализовать http обработчик для эпиков

h0ttab committed:
Реализовать http обработчик для подзадач

h0ttab committed:
Реализовать http обработчик для приоритетных задач

h0ttab committed:
Реализовать http обработчик для истории просмотров

h0ttab committed:
Рефакторинг

h0ttab committed:
Code style

h0ttab committed:
Добавить обработку пути /epics/{id}/subtasks

h0ttab committed:
Cleanup

h0ttab committed:
Cleanup

h0ttab committed:
Добавить новое исключение для случаев, когда id подзадачи совпадает с id связанного эпика

h0ttab committed:
Обновление и корректировка старых тестов и демо сценариев

h0ttab committed:
Code style

h0ttab committed:
refactor(api server): убрать static поля, сделать дефолт конструктор

h0ttab committed:
feat(api test): create test server/client as static constants

h0ttab committed:
feat(api test): extract common server and client setup/shutdown logic into an abstract class

h0ttab committed:
feat(api test): extract common CRUD test methods into an interface

h0ttab committed:
fix(api test): fix "IllegalStateException: server in wrong state Exception"

The exception caused by duplicated HttpTaskServer default constructor call

h0ttab committed:
feat(api test): add /history endpoint tests

h0ttab committed:
feat(api test): add /prioritized endpoint tests

h0ttab committed:
feat(api test): add taskId field to response message

h0ttab committed:
refactor(api test): Use independent server/client instances in test

Using independent server/client instances in tests improves environment predictability and code's readability compared to using static final values

h0ttab committed:
refactor(api test): updated methods signatures

h0ttab committed:
refactor(api test): created more specific packages for api-related entities

h0ttab committed:
feat(api test): create Task/SubTask deserializers that correctly process tasks w/o startTime field

h0ttab committed:
fix(api test): fix not auto-closing input stream after reading request body

h0ttab committed:
fix(api test): Catch more general JsonParseException than JsonSyntaxException

h0ttab committed:
fix(api test): Add proper serialization of uninitialized startTime field

h0ttab committed:
fix(api test): Consider the possibility of fields "startTime" and "id" to be null or absent when deserialize

h0ttab committed:
feat(api test): Create /tasks endpoint tests set

h0ttab committed:
refactor(api test): Move test classes to corresponding packages

h0ttab committed:
fix(api test): Fix SubTaskDeserializer crashing from null values of startTime or id fields

h0ttab committed:
feat(api test): Create /subtasks tests set

h0ttab committed:
refactor(api test): Update method signatures by adding currently thrown exceptions

h0ttab committed:
refactor(api test): Replace incorrect task type

h0ttab committed:
feat(api test): Add "500 Internal server error" template response

h0ttab committed:
feat(api test): Create /epics tests set

h0ttab committed:
style(api test): Reformat code

h0ttab committed:
refactor: Remove demo

h0ttab committed:
build(jar): Generat ArtifactManager

h0ttab committed:
docs(readme): README.md

h0ttab committed:
style(translation): Translate all service messages and responses into English

h0ttab committed:
style(http server): Add missing word

h0ttab committed:
style(http server): Add missing word

h0ttab committed:
style: Make minor style and visual improvements

h0ttab committed:
style: Modifiers order fix

h0ttab committed:
docs(readme): Text improvements

jeggga commented on src/service/FileBackedTaskManager.java:
```diff
@@ -123,7 +97,7 @@ public void loadEpic(String[] data) {
     }
 
-    public void loadSubTask(String[] data) {
+    public void loadSubTask(String[] data) throws NotFoundException {
```
Лишний проброс исключений

Данную конструкцию стоит использовать, если нет необходимости писать try-catch в текущем методе

В принципе и тут и в других местах это лишнее

h0ttab replied:
Внёс исправления:
- Если при загрузке подзадачи из CSV у неё указан некорректный эпик - эта задача просто пропускается и не выбрасывает исключение
- Заменил тип возвращаемого значения при получении эпика по id на Optional, что позволило значительно уменьшить количество выбрасываемых исключений

jeggga commented on src/service/TaskManager.java:
```diff
@@ -13,45 +14,47 @@ public interface TaskManager {
     List<SubTask> getSubTasks();
 
-    void removeAllTasks();
+    void removeAllTasks() throws NotFoundException;
```
В интерфейсе так же лишнее

h0ttab replied:
Вот тут действительно было несколько лишних "throws", спасибо что заметил. Убрал их в свежем коммите.

h0ttab committed:
refactor(task manager): Remove unnecessary "throws" statements

h0ttab committed:
refactor(epic): Optimize epic update algorithm

Since the only fields of epic that may be updated are title and description, it's more efficient to simply update these values rather than actually transfer all data to a new epic.

h0ttab committed:
feat(task manager): Skip invalid subtask instead of throwing an exception

h0ttab committed:
refactor: Use Optional<T> instead of throwing unnecessary exceptions

h0ttab committed:
refactor: Use Optional<T> instead of throwing unnecessary exceptions

h0ttab committed:
fix(tests): Update tests for use of Optional<T> than throwing exceptions

h0ttab committed:
style: Fix code style issues
